File: Lexer/Token.cpp
/**
 * @file Token.hpp
 * @author 35921
 * @date 2026/2/4
 * @brief 词法分析器的Token（标记）相关实现
 *
 * 该文件实现了Token类以及将TokenType转换为字符串的工具函数，
 * 用于词法分析阶段对源代码进行标记化处理。
 */

#ifndef TOKEN_CPP
#define TOKEN_CPP

#include "Token.hpp"
#include <string>

namespace lexer::token {

    std::string tokenTypeToString(const TokenType type) {
        switch (type) {
        case TokenType::IDENTIFIER:
            return "IDENTIFIER";
        case TokenType::KEYWORD:
            return "KEYWORD";
        case TokenType::NUMBER:
            return "NUMBER";
        case TokenType::STRING:
            return "STRING";
        case TokenType::FORMAT_STRING:
            return "FORMAT_STRING";
        case TokenType::OPERATOR:
            return "OPERATOR";
        case TokenType::END_OF_LINE:
            return "END_OF_LINE";
        case TokenType::END_OF_FILE:
            return "END_OF_FILE";
        default:
        case TokenType::UNKNOWN:
            return "UNKNOWN";
        }
    }

} // namespace lexer::token
#endif // TOKEN_CPP

File: src/token.rs
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum TokenType {
    Identifier,
    Keyword,
    Number,
    String,
    FormatString,
    Operator,
    EndOfLine,
    EndOfFile,
    Unknown,
}

impl fmt::Display for TokenType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            TokenType::Identifier => "identifier",
            TokenType::Keyword => "keyword",
            TokenType::Number => "number",
            TokenType::String => "string",
            TokenType::FormatString => "format_string",
            TokenType::Operator => "operator",
            TokenType::EndOfLine => "end_of_line",
            TokenType::EndOfFile => "end_of_file",
            TokenType::Unknown => "unknown",
        };
        write!(f, "{}", s)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub r#type: TokenType,
    pub value: String,
    pub line: i32,
    pub col: i32,
}

impl Token {
    pub fn new(r#type: TokenType, value: impl Into<String>) -> Self {
        Token {
            r#type,
            value: value.into(),
            line: 0,
            col: 0,
        }
    }

    pub fn with_pos(r#type: TokenType, value: impl Into<String>, line: i32, col: i32) -> Self {
        Token {
            r#type,
            value: value.into(),
            line,
            col,
        }
    }
}

#[allow(dead_code)]
#[deprecated(note = "TokenType can now be turned into string inline")]
fn token_type_to_str(r#type: &TokenType) -> &str {
    match r#type {
        TokenType::Identifier => "identifier",
        TokenType::Keyword => "keyword",
        TokenType::Number => "number",
        TokenType::String => "string",
        TokenType::FormatString => "format_string",
        TokenType::Operator => "operator",
        TokenType::EndOfLine => "end_of_line",
        TokenType::EndOfFile => "end_of_file",
        TokenType::Unknown => "unknown",
    }
}
